doc: add sign-action.md (#5)
notation-playground/notation-integration-with-ACR-and-AKV: sign-action.md

# Sign an image with Notation in GitHub Actions

This document walks you through how to create a GitHub Actions workflow to achieve the following goals:

1. Build an image and push it to Azure Container Registry (ACR)
2. Sign the image with Notation and Notation AKV plugin with a signing key stored in Azure Key Vault (AKV). The generated signature is automatically pushed to ACR

## Prerequisites

- You have created a Key Vault in AKV and created a self-signed signing key and certificate. You can follow this [doc](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-tutorial-sign-build-push#create-a-self-signed-certificate-azure-cli) to create self-signed key and certificate for testing purposes. 
- You have created a registry in Azure Container Registry
- You have a GitHub repository to store the sample workflow file and GitHub Secrets

## Create GitHub Secrets to store credentials

> Enter your GitHub repository, create two encrypted secrets `ACR_PASSWORD` and `AZURE_CREDENTIALS` for the repository to authenticate with ACR and AKV. See the [GitHub doc](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) for details.

Reviewer: By reading this, I think we still need a brief definition of these two terms. Btw, we need to write out what the `GitHub doc` link is about, otherwise, it will likely be skipped by the reader.
```suggestion
Enter your GitHub repository, create two encrypted secrets `ACR_PASSWORD` and `AZURE_CREDENTIALS` in the repository to authenticate with ACR and AKV. See [creating encrypted secrets for a repository](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) for details.

- `ACR_PASSWORD`: the password of your ACR registry where your image will be stored.
- `AZURE_CREDENTIALS`: the credential to access your AKV where your signing key pair is stored.
```

> ### Create a GitHub Secret `ACR_PASSWORD`
> 
> Create the first GitHub Secret naming `ACR_PASSWORD`, enter the password of your ACR registry where your image will be stored in the GitHub Secret value.

Reviewer: Since the whole session is under title `Create GitHub Secrets to store credentials`, we don't need to add `GitHub Secret` again in subtitles, it's kind of redundant and overwhelming to have the same word appear too many times in the doc.
Given the above changes, now we can be very brief and straightforward, i.e., directly and only provide actions that a user needs to follow under the sub-session.
```suggestion
### Create `ACR_PASSWORD`

Create a new GitHub repository secret. Type `ACR_PASSWORD` in the `Name` field. Enter your ACR password into the `Secret` field.
```

### Create a GitHub Secret `AZURE_CREDENTIALS`

Follow the steps below to get the value of `AZURE_CREDENTIALS`. This is credential to access your AKV where your signing key pair is stored.

- Execute the following commands to create a new service principal on Azure. 

```
# login using your own account
az login

# Create a service principal
spn=notationtest
az ad sp create-for-rbac -n $spn --sdk-auth
```

> [!IMPORTANT]

> > 1. Copy the entire JSON output from the `az ad sp` command execution result into the value of GitHub Secret `AZURE_CREDENTIALS`. See [this doc](https://learn.microsoft.com/en-us/azure/developer/github/connect-from-azure?tabs=azure-portal%2Cwindows#add-the-service-principal-as-a-github-secret) for reference.

Reviewer: At this point, the reader already knows that `AZURE_CREDENTIALS` is a GitHub Secret. We don't need to repeat it everytime.
```suggestion
> 1. Copy the entire JSON output from the `az ad sp` command execution result into the `Secret` field of `AZURE_CREDENTIALS`. See [this doc](https://learn.microsoft.com/en-us/azure/developer/github/connect-from-azure?tabs=azure-portal%2Cwindows#add-the-service-principal-as-a-github-secret) for reference.
```

>
> 2. Save the `clientId` from the JSON output into an environment variable (without double quotes) as it will be needed in the next step:
>```
>  clientId=<clientId_from_JSON_output_of_last_step>
>```

- Grant the AKV access permissions to the service principal that we created in the previous step.

```
# set policy for your AKV
akv=<your_akv_name>
az keyvault set-policy --name $akv --spn $clientId --certificate-permissions get --key-permissions sign --secret-permissions get
```

See [az keyvault set-policy](https://learn.microsoft.com/en-us/cli/azure/keyvault?view=azure-cli-latest#az-keyvault-set-policy) for reference.

### Create the GitHub Actions workflow

> - Create a `.github/workflows` directory in your repository on GitHub if this directory does not already exist.

Reviewer: Just an open discussion, there are users who would do a `git clone`, do their work, `git add`, `git commit`, `git push`. Although this might not be the recommended way to do things, this is definitely the most popular way among developers, and these people are the main readers of this doc.
```suggestion
- Create a `.github/workflows` directory in your GitHub repository if this directory does not already exist.
```
(a `GitHub repository` is not necessary on `GitHub`).

> - In the `.github/workflows directory`, create a file named `<your_workflow>.yml`. For more information, see [Creating new files](https://docs.github.com/en/repositories/working-with-files/managing-files/creating-new-files).

Reviewer: Same here. 
```suggestion
- In the `.github/workflows` directory, create a file named `<your_workflow>.yml`.
```

- You can copy the [signing template workflow](https://github.com/notation-playground/notation-integration-with-ACR-and-AKV/blob/template/sign-template.yml) from the collapsed section below into your own `<your_workflow>.yml` file.

> - Update the environmental variables based on your environment by following the comments in the template. Save it and commit it to the repository.

Reviewer: ```suggestion
- Update the environmental variables based on your environment by following the comments in the template. Save and commit it to the repository.
```

# Sign an artifact with Notation in GitHub Actions

This document walks you through how to create a GitHub Actions workflow to sign the latest build of the project with the Notation Azure Key Vault plugin, and push the signed image with its associated signature to Azure Container Registry.

### Prerequisites

> - You have created a Key Vault in Azure Key Vault and created a self-signed signing key and certificate. You can follow this [doc](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-tutorial-sign-build-push#create-a-self-signed-certificate-azure-cli) to create self-signed key and certificate for testing purposes

Author: This doc is for quick-start purposes. We can point users to another doc of signing with a CA cert (certificate chain) until this [PR ](https://github.com/MicrosoftDocs/azure-docs-pr/pull/249166/files#diff-ecfd75bf480181ab17ab74f41f7093a0f471bb6c31c3a59d001d36626c603232) is merged.

>
> 2. Save the `clientId` from the JSON output into an environment variable (without double quotes) as it will be needed in the next step:
>```
>  clientId=<clientId_from_JSON_output_of_last_step>
>```

### Create GitHub Secret to store credentials 

- Create an encrypted secret to store the Azure Credentials in your own GitHub repository. See [GitHub Docs](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) for details.

- Add the JSON output of the following `az ad sp` command to the value of GitHub Secret. Naming the GitHub Secret as `AZURE_CREDENTIALS`. 

- Similarly, create another encrypted secret to store the registry credential or password in your own GitHub repository. For example, we use `ACR_PASSWORD` to store the password of the ACR registry in the GitHub Repository secret.

### Grant AKV permission to the service principal

Grant AKV permission to the service principal that we created in the previous step.

```
# set policy for your AKV
akv=<your_akv_name>
az keyvault set-policy --name $akv --spn $clientId --certificate-permissions get --key-permissions sign --secret-permissions get
```

> See [az keyvault set-policy](https://learn.microsoft.com/en-us/cli/azure/keyvault?view=azure-cli-latest#az-keyvault-set-policy) for details.

Reviewer: I think this part should be included under the `prerequisites` session's `AZURE_CREDENTIALS` section.

Reviewer: Based on our discussion, we will move line18 to line 57 under `AZURE_CREDENTIALS`.

### Edit the sample workflow

> - Create a `workflows` folder under `.github` and create a `workflow.yml` to run and test the CI/CD pipeline. You can copy the [signing template workflow](https://github.com/notation-playground/notation-integration-with-ACR-and-AKV/blob/template/sign-template.yml) to your own `workflow.yml` file. 

Reviewer: ```suggestion
- Create `<your_repo>/.github/workflows/<your_workflow>.yml` to run and test the CI/CD pipeline. You can copy the [signing template workflow](https://github.com/notation-playground/notation-integration-with-ACR-and-AKV/blob/template/sign-template.yml) to your own `<your_workflow>.yml` file. 
```

Reviewer: Based on our discussion, we will include the contents of the `signing template workflow` directly in this doc.
